### PR TITLE
Fix: no-constant-condition doesn't introspect arrays (fixes #12225)

### LINKED ISSUE #12225: no-constant-condition false positive on string compare
[Check out Demo Page for more details](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tY29uc3RhbnQtY29uZGl0aW9uOiBcImVycm9yXCIqL1xuZnVuY3Rpb24gaXNEYXRlKHR5LCB0bSwgdGQpIHtcbiAgY29uc3QgW3ksIG0sIGRdID0gbmV3IERhdGUoKS50b0lTT1N0cmluZygpLnNwbGl0KC9bLVRdLykubWFwKHggPT4gK3gpO1xuICBpZiAoJycgKyBbeSwgbSwgZF0gPT09ICcnICsgW3R5LCB0bSwgdGRdKSB7XG4gICAgcmV0dXJuIHRydWU7XG4gIH1cbiAgcmV0dXJuIGZhbHNlO1xufVxuY29uc29sZS5sb2coaXNEYXRlKDIwMTksOSw1KSk7XG5cbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtY2F0Y2giOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJuby13aXRoIjoyLCJyZXF1aXJlLWF0b21pYy11cGRhdGVzIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Mn0sImVudiI6eyJicm93c2VyIjp0cnVlfX19)

**Tell us about your environment**

* **ESLint Version:** 6.3.0

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
    "parserOptions": {
        "ecmaVersion": 11,
        "sourceType": "script",
        "ecmaFeatures": {}
    },
    "rules": {
        "constructor-super": 2,
        "for-direction": 2,
        "getter-return": 2,
        "no-async-promise-executor": 2,
        "no-case-declarations": 2,
        "no-class-assign": 2,
        "no-compare-neg-zero": 2,
        "no-cond-assign": 2,
        "no-const-assign": 2,
        "no-constant-condition": 2,
        "no-control-regex": 2,
        "no-debugger": 2,
        "no-delete-var": 2,
        "no-dupe-args": 2,
        "no-dupe-class-members": 2,
        "no-dupe-keys": 2,
        "no-duplicate-case": 2,
        "no-empty": 2,
        "no-empty-character-class": 2,
        "no-empty-pattern": 2,
        "no-ex-assign": 2,
        "no-extra-boolean-cast": 2,
        "no-extra-semi": 2,
        "no-fallthrough": 2,
        "no-func-assign": 2,
        "no-global-assign": 2,
        "no-inner-declarations": 2,
        "no-invalid-regexp": 2,
        "no-irregular-whitespace": 2,
        "no-misleading-character-class": 2,
        "no-mixed-spaces-and-tabs": 2,
        "no-new-symbol": 2,
        "no-obj-calls": 2,
        "no-octal": 2,
        "no-prototype-builtins": 2,
        "no-redeclare": 2,
        "no-regex-spaces": 2,
        "no-self-assign": 2,
        "no-shadow-restricted-names": 2,
        "no-sparse-arrays": 2,
        "no-this-before-super": 2,
        "no-undef": 2,
        "no-unexpected-multiline": 2,
        "no-unreachable": 2,
        "no-unsafe-finally": 2,
        "no-unsafe-negation": 2,
        "no-unused-labels": 2,
        "no-unused-vars": 2,
        "no-useless-catch": 2,
        "no-useless-escape": 2,
        "no-with": 2,
        "require-atomic-updates": 2,
        "require-yield": 2,
        "use-isnan": 2,
        "valid-typeof": 2
    },
    "env": {
        "browser": true
    }
}
```
</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/*eslint no-constant-condition: "error"*/
function isDate(ty, tm, td) {
  const [y, m, d] = new Date().toISOString().split(/[-T]/).map(x => +x);
  if ('' + [y, m, d] === '' + [ty, tm, td]) {
    return true;
  }
  return false;
}
console.log(isDate(2019,9,5));
```

**What did you expect to happen?**
No error

**What actually happened? Please include the actual, raw output from ESLint.**
> 4:7 - Unexpected constant condition. (no-constant-condition)

### PR DESCRIPTION
**Pull request by tickct**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

https://github.com/eslint/eslint/issues/12225

**What changes did you make? (Give an overview)**

no-constant-condition uses a recursive function `isConstant` to detect if items are constants. 

Previously we were presuming that all arrays were always constant. Since an array can hold variables, if we are inside of a binary evaluation it is possible for the array to change and there for change the conditional.   

Now, if we are in a binary expression, we will calculate isConstant on each element of the array, and only return true if all elements are constant

**Is there anything you'd like reviewers to focus on?**

Per the espree docs, array elements of sparse arrays have values of null. I have added a check at the beginning of the function to handle this, but I could see the argument to handle this situation in other ways such as a filter.

### CODE REVIEW

**Comment by mdjermanovic** [comment]
Thanks for the PR! Marking as accepted as the issue is accepted.

The solution seems correct for the #12225 false positive, but I think that it could also introduce some new false negatives? For example:

```js
/*eslint no-constant-condition: "error"*/

if ([a, b] === [c, d]) {
  // ...
}
```

**Comment by mdjermanovic** [comment]
> Per the espree docs, array elements of sparse arrays have values of null. I have added a check at the beginning of the function to handle this, but I could see the argument to handle this situation in other ways such as a filter.

It's okay to check at the beginning, e.g., `no-self-assign` handles this in a similar way.

Though, the espree bug comment can be removed, as it isn't a bug :)

**Comment by tickct** [comment]
This does create some false positives as you noted, however trying to be exhaustive in the permutations seem unfeasible, as we would need an list of each binary operator and wether or not that operator is valid. I can either close the scope, allowing for only `+` operators, which would fulfill this bug, but may cause the issue to reopen in the future if for some reason someone is using `-` for some reason, or I can make only direct equality operators `===,==,!==,!=` fail and possibly give us other false negatives and deal with those as I come up.

I am not sure if we would rather have more false negatives, or more false positives. Let me know what you guys would prefer

**Comment by mdjermanovic** [comment]
When looking at these particular false positives and negatives, supporting the use case of comparing arrays by their string presentations at the cost of missing to report cases like `[a, b] === [c, d]` doesn't seem like a good thing to me, personally. Or, at the cost of increasing the complexity too much.

There is also the following workaround: `String([foo])` instead of `'' + [foo]` is valid by this rule.

I'm not sure is there another use case with 'dynamic' arrays that could currently produce a false positive? Other than converting the array to string at some point.

I'd vote to limit the scope and allow just `string_literal + [dynamic_array]` in this PR, but better wait for other opinions.

**Comment by tickct** [comment]
Since no one else has weighed in and I agree with @mdjermanovic, I have update the PR to only support the narrow scope of being a child of a `+` operator.

Noted this does now allow for `if('' + [a]){}`,  but in testing `""+[undefined]` evaluates to `""` so it is indeed possible for the condition to evaluate to both truthy and falsy.

File: lib/rules/no-constant-condition.js [hunk 1]
```diff
@@ -90,14 +90,25 @@ module.exports = {
...
                 case "FunctionExpression":
                 case "ObjectExpression":
-                case "ArrayExpression":
                     return true;
 
+                case "ArrayExpression": {
+                    if (!(node.parent.operator === "+")) {
+                        return true;
+                    }
+                    return node.elements.every(element => isConstant(element, false));
```
Comment (mdjermanovic) [inline lib/rules/no-constant-condition.js]: Perhaps we should also check the parent's type and continue only if it's `BinaryExpression`? I think that this at the moment also allows unary `+` which almost always evaluates to `NaN` (except when the array has 1 element, but then the array is unnecessary anyway).

I was also thinking to allow `+` on an array only if the other operand is a string literal, to minimize the change in behavior, but I'm not sure about that.

**Comment by kaicataldo** [comment]
I agree that we should scope this bug fix to the accepted bug only, and leave any further changes to a separate proposal. @tickct Do you mind signing the CLA?

**Comment by kaicataldo** [comment]
@tickct Friendly ping

**Comment by tickct** [comment]
I have signed the CLA, I unfortunately had not configured git with my email until after I did the initial push, so now it is showing two authors one of which has not signed, I am not sure if there is a way to rectify that.

**Comment by kaicataldo** [comment]
@tickct Understood. You should be able to follow the instructions [here](https://stackoverflow.com/questions/3042437/how-to-change-the-commit-author-for-one-specific-commit/28845565). Thanks!

File: tests/lib/rules/no-constant-condition.js [hunk 2]
```diff
@@ -73,6 +73,14 @@ ruleTester.run("no-constant-condition", rule, {
...
         "if ((foo || {}) instanceof obj) {}",
 
+        // #12225
+        "if ('' + [y] === '' + [ty]) {}",
+        "if ('a' === '' + [ty]) {}",
+        "if ('' + [y, m, d] === 'a') {}",
+        "if ('' + [y, 'm'] === '' + [ty, 'tm']) {}",
+        "if ('' + [y, 'm'] === '' + ['ty']) {}",
+        "if ([,] in\n\n($2))\n ;\nelse\n ;",
+
```
Comment (mdjermanovic) [inline tests/lib/rules/no-constant-condition.js]: Might be nice to have a test with spread element in an array.

File: tests/lib/rules/no-constant-condition.js [hunk 3]
```diff
@@ -183,6 +192,40 @@ ruleTester.run("no-constant-condition", rule, {
...
+        {
+            code: "if([a]==[a]) {}",
+            errors: [{ messageId: "unexpected", type: "BinaryExpression" }]
+        },
+        {
+            code: "if([a] - '') {}",
+            errors: [{ messageId: "unexpected", type: "BinaryExpression" }]
+        },
+        {
+            code: "if(+1) {}",
```
Comment (mdjermanovic) [inline tests/lib/rules/no-constant-condition.js]: I think this was meant to be `+[a]`? (a test to ensure that the change in this PR applies only to binary `+`)

**Comment by kaicataldo** [comment]
> I think we can merge this PR as is, and add a test later.

For clarity, are you referring to https://github.com/eslint/eslint/pull/12307#discussion_r361823367?

**Comment by mdjermanovic** [comment]
> > I think we can merge this PR as is, and add a test later.
> 
> For clarity, are you referring to [#12307 (comment)](https://github.com/eslint/eslint/pull/12307#discussion_r361823367)?

Yes. 

This PR has been open for a long time because it wasn't clear what should be done to fix the related issue and not produce new false negatives at the same time.

The scope of the change looks good to me now, the code which implements it also looks good and the change is well-tested, so I think that one nice-to-have test shouldn't block this. I can add the test later.

**Comment by kaicataldo** [comment]
@mdjermanovic Sounds good :+1:

@tickct Apologies for the delay, and thank you contributing to ESLint!

**Pull request merged**